### ISSUE TITLE
Damage roulette is buggy

### ISSUE THREAD

**Issue by afwbkbc** [We-the-People-civ4col-mod/Mod#177]
Yesterday I tried to attack enemy Galleon ( strength 36 ) with Privateer ( strength 33 ), and Privateer died while dealing zero damage to Galleon.
So 10% strength difference resulted in 100% defeat ( no damage dealt to Galleon at all ).
I wonder if it's just me being unlucky or some kind of bug?

**Comment by ErikSandbraaten** [comment]
While the combat system can surely be debated, it is not buggy just because you dislike the outcome. Do you remember all the times that you won against all odds ? Funny how people don't complain about that!

**Comment by Nightinggale** [comment]
`
While the combat system can surely be debated, it is not buggy just because you dislike the outcome. Do you remember all the times that you won against all odds ? Funny how people don't complain about that!
`
Is that Erik or Sid Meier talking? You get more or less the same from the link. Sid Meier talks about how players think it's unfair if they lose when the AI has 1:3 chance of winning while they think it's ok to win with 1:3 chance because they are good and can defeat the odds or something like that.

I agree the outcome of this combat isn't a bug. I view it more like disagreement with the combat system design.

**Comment by afwbkbc** [comment]
"This means if you are unlucky enough to lose every single round, then the enemy will stay unharmed." - so how many rounds did i lose there to do zero damage with str 33 vs str 36?
"Do you remember all the times that you won against all odds ?" - nope, not even once.

**Comment by ShadesOT** [comment]
Let's put rewriting combat logic on the todo list. :-) I would like it.

>  so how many rounds did i lose there to do zero damage with str 33 vs str 36?

You actually can look that up in the message log under the combat tab.

**Comment by MichaelQuell** [comment]
I think a draw should be far more common in this game.

**Comment by ShadesOT** [comment]
What happened to "fight to the death"?

**Comment by Nightinggale** [comment]
`
What happened to "fight to the death"?
`
It died because it failed to accept a draw when it was overpowered.

There is a good point to a draw. Damaged units can then flee back to the colonies to heal. This means it can be beneficial to use troops, which have sacrificed combat power for speed as they can be used to chase such damaged units.

**Comment by LibSpit** [comment]
Draws come from the withdrawal chance 'skill'. The ability to 'not lose' in a bad outcome.

So if you wanted things to 'draw' more often then you just apply withdrawal chance amounts as a default.

**Comment by NoFreeUsernameDammit** [comment]
Draw comes from two things: withdraw ability and running out of combat rounds.

The combat is turn-based, and number of rounds is, in vanilla Colonization, unlimited-meaning combat will go on until death, unless attacker withdraws. RaR changed rounds to limited, so now the units often run out of combat rounds. 

Withdraw chance applies only for attacker, and I think it's rolled every time attacker loses a round. The idea in original design was to allow some units to act as skirmishers, weakening the enemy and retreating.

**Comment by LibSpit** [comment]
I may be thinking of a civ4 mod that changed withdrawal to work for defenders as well.

**Comment by ErikSandbraaten** [comment]
In case anyone would like to experiment with the combat calculations, have a look @ CvUnit::resolveCombat(...)
Currently the number of combat rounds are hard-coded to 7. Arguable, this should be a XML parameter.
Note that if you change it, you'll also have to change the following constant "int iMaxCombatRounds = 7" in  ::getCombatOddsDraw(...) to make the combat odds calculation remain correct.

**Comment by orlanth** [comment]
Nice, moving in the direction of getting more combat variables exposed to XML will be great for moddability. 👍 There is plenty of scope for making vanilla combat more interesting, by promotions enabling situational modifiers that can make combat more than a generic contest of “strength value vs strength value +/- randomness”.

Rather than a single generic value for “strength”, some mods like FfH2 have enabled attack strength and defense strength values that can be set independently. In many cases I liked this a lot, but I think many modders may not to deal with sepaparte attack/defense values for all units and some may not want to use this at all. Perhaps a simple way to enable moddable attack/defense advantages would be promotioninfos <iAttackModifier> and <iDefenseModifier> tags that can provide % strength modifiers when attacking or defending.

I agree more “draws” / partial defeats could be nice in enabling more nuanced battle outcomes and and result in less player frustration from outright losses; however as Night said this should be balanced with avoiding frustration from constantly chasing retreating “unkillable” units. To enable moddability of withdrawal for attacking or defending units, <iWithdrawalAttack> and <iWithdrawalDefense> tags in Promotioninfos could be added. To enable some units to be better at chasing down evasive prey, <iPursuit> could be subtracted from the Withdrawal chance of their retreating opponent.

I can’t recall what was discussed about unit capture, but an <iCapture> chance that allows a victorious unit to capture an opponent that failed to retreat could also be very cool.

M:C had also brought back the option of “First Strike” attacks from civ4bts, another nice option for adding strategy and flavor to vanilla combat.

**Comment by LibSpit** [comment]
yeah in FTTW we added a ranged strength stat, and we had attack and defense modifiers this gave us the ability to have melee/ranged strengths and an ability to have very strong attacks and weak defensive, health etc.

**Comment by Nightinggale** [comment]
We technically have 4 combat cases. Attack and defend and in both cases the unit can both try to hit and avoid being hit. Think pikemen vs cavalry. The pikemen are slow and can't be aggressive against cavalry, hence not good at attacking. However if cavalry try to charge pikemen, all they meet is a wall of spikes and they will be slaughtered. This would give pikemen a really high change to hit and damage vs mounted units, but only when the mounted unit attack.

I would like to add that #127 brings up the concept of units having a list of UnitCombatTypes rather than just one. This means a unit can end up with both gunpowder and mounted, accepting promotions for both and the enemy can use +% combat strength vs either (though should they stack or just best for each promotion? Something to consider).

We can also consider two types of armor: hit avoidance and damage reduction. An unarmored unit can be very agile in moving out of where the enemy hits while a heavily armored unit might not be able to move around, but it takes less damage each time it's hit because the armor does the job.

`
M:C had also brought back the option of “First Strike” attacks from civ4bts, another nice option for adding some strategy and flavor beyond “strength” values to vanilla combat.
`
The idea is interesting, though I'm not happy with the implementation. We should keep it simple. Just have units have an int of number of first strikes. During combat each unit gets a counter of number of first strikes from the enemy and if it's > 0, subtract one and skip the turn. It's much simpler for the code and more intuitive. Mixing it with armor type has made the code.... rather complex and less intuitive to the player.

**Comment by CommanderBello** [comment]
Currently the number of combat rounds are hard-coded to 7. Arguable, this should be a XML parameter.
------
Hello guys, a new contributor here. Please forgive me my poor English, as I am not a native English speaker.
As the above mentioned number of combat rounds was introduced to RaR based on my modmod at that time, I would like to point out some things.
The number "7" was based on experience and chances to have some draws in combat while still giving either side the chance to have a win. At that time I was (and still am) quite inexperienced in programming, so for me it was the easiest way to have that number hardcoded. 
I agree wholeheartedly that having it in XML would be a much better way. 
(Will continue this later as some RL action is requested right now)

**Comment by ErikSandbraaten** [comment]
I am working with f1rpo to overhaul the civ4 combat mechanics: https://forums.civfanatics.com/threads/advanced-civ.614217/page-9 (the discussion started on page-2)

Anyways, I am mentioning it here since when completed some or parts of it could be ported to WTP.

@CommanderBello :  Your limited combat rounds mechanics will be included :)
